xlators/cluster/dht/src/dht-helper.c

```diff
@@ -1792,18 +1792,30 @@ dht_inode_ctx_layout_set(inode_t *inode, xlator_t *this,
 {
     dht_inode_ctx_t *ctx = NULL;
     int ret = -1;
+    uint64_t ctx_int = 0;
 
-    ret = dht_inode_ctx_get(inode, this, &ctx);
-    if (!ret && ctx) {
-        ctx->layout = layout_int;
-    } else {
-        ctx = GF_CALLOC(1, sizeof(*ctx), gf_dht_mt_inode_ctx_t);
-        if (!ctx)
-            return ret;
-        ctx->layout = layout_int;
-    }
+    LOCK(&inode->lock);
+    {
+        ret = __inode_ctx_get(inode, this, &ctx_int);
+        if (!ret) {
+            ctx = (dht_inode_ctx_t *)(uintptr_t)ctx_int;
+            if (ctx) {
+                ctx->layout = layout_int;
```

Comment:
I would take the layout reference inside this function. Callers shouldn't care about it.

Reply:
We are calling this function during calling of layout_present so we don't take reference in this case.

Reply:
Why not ? who's taking the reference then ? we are releasing the reference when the layout is removed from the context, shouldn't we take the reference when it's added ?

Reply:
The reference is taking while creating a layout, i mean while it call dht_layout_new.Initially the layout refcnt is 1.

Reply:
I don't fully understand this. `dht_layout_preset()` calls `dht_layout_for_subvol()` to get the layout, but that function seems to use a pre-created layout, and apparently it can be used more than once (another `dht_layout_preset()` call may use the same layout). How does this work without taking additional references ?